centralstore/redis_store.go

```diff
@@ -762,84 +761,61 @@ func (t *tracesStore) getTraceStatuses(ctx context.Context, client redis.Client,
-	wgFans.Add(1)
-	go func() {
-		defer wgFans.Done()
-		for status := range faninChan {
-			statuses[status.TraceID] = status
-		}
-	}()
-
-	// Randomly select the number of goroutines to use, using most of what we've been given
-	// for our maximum number of connections in the redis pool.
-	// It seems (after messing around) that adding a bit of randomness here helps with the performance.
-	maxGoroutines := t.config.GetRedisMaxActive()
-	if maxGoroutines <= 0 {
-		maxGoroutines = 1
+	// First, calculate the max number of goroutines to use (leaving a few for the rest of the world).
+	numGoroutines := t.config.GetRedisMaxActive() - 5
+	// but don't use more than 1 for every 10 traceIDs
+	if numGoroutines > len(traceIDs)/10 {
+		numGoroutines = len(traceIDs) / 10
+	}
+	if numGoroutines <= 0 {
+		numGoroutines = 1
 	}
-	randRange := maxGoroutines/4 + 1
-	numGoroutines := rand.Intn(randRange) + maxGoroutines - randRange
 	otelutil.AddSpanField(statusSpan, "num_goroutines", numGoroutines)
 
-	// create workers to get the traceIDs and send their statuses to the fanin channel
-	// They will pull from the fanout channel and terminate when it closes
-	wgWorkers := sync.WaitGroup{}
-	for i := 0; i < numGoroutines; i++ {
-		wgWorkers.Add(1)
-		go func(i int) {
-			conn := client.Get()
-			defer conn.Close()
-			queries := 0
-			found := 0
-			_, span := otelutil.StartSpanWith(ctx, t.tracer, "getTraceStatusesWorker", "worker", i)
-			defer span.End()
-			defer wgWorkers.Done()
-			for traceID := range fanoutChan {
-				status := &centralTraceStatusRedis{}
-				err := conn.GetStructHash(t.traceStatusKey(traceID), status)
-				queries++
-				if err != nil {
-					if errors.Is(err, redis.ErrKeyNotFound) {
-						continue
-					}
-					statusSpan.RecordError(err)
-					continue
+	// Now create a worker factory that will create a worker that will get the status for a traceID.
+	// We also want telemetry so we do a little bit of prework and a cleanup function.
+	workerFactory := func(i int) (func(traceID string) *CentralTraceStatus, func(int)) {
+		conn := client.Get()
+		queries := 0
+		found := 0
+		_, span := otelutil.StartSpanWith(ctx, t.tracer, "getTraceStatusesWorker", "worker", i)
+		worker := func(traceID string) *CentralTraceStatus {
+			status := &centralTraceStatusRedis{}
+			err := conn.GetStructHash(t.traceStatusKey(traceID), status)
+			queries++
+			if err != nil {
+				if errors.Is(err, redis.ErrKeyNotFound) {
+					return nil
 				}
-				faninChan <- normalizeCentralTraceStatusRedis(status)
-				found++
+				statusSpan.RecordError(err)
+				return nil
 			}
+			found++
+			return normalizeCentralTraceStatusRedis(status)
+		}
+		cleanup := func(i int) {
+			conn.Close()
 			otelutil.AddSpanFields(span, map[string]interface{}{
 				"num_queries": queries,
 				"num_found":   found,
 			})
-		}(i)
+			span.End()
+		}
+		return worker, cleanup
+	}
+	// this filters out the nil statuses
+	predicate := func(s *CentralTraceStatus) bool {
```

Comment:
I don't see the previous implementation filtering nil statuses, what is the impact of dropping these? Is a nil status an error or does it mean redis didnt have that trace id anymore?

Reply:
The previous implementation did it differently -- the worker had a loop that it `continue`d if the result was meaningless. In this one, it adds nil and we filter it outside the worker.